### PR TITLE
Allow iOS and macOS plugins to share darwin directory

### LINKED ISSUE #115085: Allow directly sharing iOS and macOS plugin native code
See previous discussion in https://github.com/flutter/flutter/issues/52285. At that time the decision was to use symlinks, but that's kind of a pain due to https://github.com/dart-lang/pub/issues/3143, which means all files have to be symlinked, rather than the directories (which wasn't true when the decision was made initially). For non-trivial plugins that can be a lot of files (e.g., `in_app_purchase`).

I'd need to look at the logic again, but I think it would be plausible to have a pubspec.yaml entry in the platform section that would allow opting into a shared directory instead of the currently hard-code `ios` or `macos`. If it's not too much complexity, I think the maintenance benefits for plugins would be worth the effort.

See also https://github.com/flutter/flutter/issues/114179

(/cc @jmagman for thoughts)

### PR DESCRIPTION
**Pull request by jmagman**
iOS and macOS plugins can share code in a `darwin` directory.

Design at https://flutter.dev/go/plugin-shared-ios-macos

Fixes https://github.com/flutter/flutter/issues/115085

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by jmagman** [comment]
This is pending a a design doc.

File: dev/devicelab/lib/tasks/plugin_tests.dart [hunk 1]
```diff
@@ -151,6 +156,79 @@ class $dartPluginClass {
...
+#elseif os(iOS)
+import Flutter
+#endif
+
+public class $pluginClass: NSObject, FlutterPlugin {
+  public static func register(with registrar: FlutterPluginRegistrar) {
+#if os(macOS)
+    let channel = FlutterMethodChannel(name: "$name", binaryMessenger: registrar.messenger)
+#elseif os(iOS)
+    let channel = FlutterMethodChannel(name: "$name", binaryMessenger: registrar.messenger())
```
Comment (stuartmorgan) [inline dev/devicelab/lib/tasks/plugin_tests.dart:218]: I really want to see if we can fix this; it bit me recently too. For Obj-C the change is transparent, but it's unclear to me if there's a way to change it without breaking Swift plugins. (Not for this PR, just venting.)

File: packages/flutter_tools/bin/podhelper.rb [hunk 2]
```diff
@@ -288,7 +293,7 @@ def flutter_parse_plugins_file(file, platform)
 
   # dependencies_hash.dig('plugins', 'ios') not available until Ruby 2.3
   return [] unless dependencies_hash.has_key?('plugins')
-  return [] unless dependencies_hash['plugins'].has_key?('ios')
```
Comment (stuartmorgan) [inline packages/flutter_tools/bin/podhelper.rb:291]: How has this been working?

File: packages/flutter_tools/lib/src/platform_plugins.dart [hunk 3]
```diff
@@ -227,15 +236,17 @@ class AndroidPlugin extends PluginPlatform implements NativeOrDartPlugin {
...
     required this.name,
     required this.classPrefix,
     this.pluginClass,
     this.dartPluginClass,
     bool? ffiPlugin,
     this.defaultPackage,
-  }) : ffiPlugin = ffiPlugin ?? false;
+    bool? sharedDarwinSource,
+  }) : ffiPlugin = ffiPlugin ?? false,
+       sharedDarwinSource = sharedDarwinSource ?? false;
```
Comment (stuartmorgan) [inline packages/flutter_tools/lib/src/platform_plugins.dart:249]: Do we actually need to be able to explicitly pass `sharedDarwinSource: null` for some reason? If not we should be able to make it `this.sharedDarwinSource = false` above and not need this I think.

File: packages/flutter_tools/test/general.shard/plugins_test.dart [hunk 4]
```diff
@@ -47,6 +48,8 @@ class _PluginPlatformInfo {
   /// The package entry for an Android plugin implementation using pluginClass.
   final String? androidPackage;
 
+  final bool? sharedDarwinSource;
```
Comment (stuartmorgan) [inline packages/flutter_tools/test/general.shard/plugins_test.dart:51]: Same here; can this be a default-false non-nullable?

**Review by stuartmorgan-g — APPROVED** [review]
LGTM with optional nits. Ship it!

File: packages/flutter_tools/lib/src/platform_plugins.dart (hunk 3 above)
Comment (jmagman) [reply packages/flutter_tools/lib/src/platform_plugins.dart:249]: It makes it easier to do:
```dart
  sharedDarwinSource: yaml[kSharedDarwinSource] as bool?,
```
You're right the property should be nonnull though even when the constructor allows null.

File: packages/flutter_tools/test/general.shard/plugins_test.dart (hunk 4 above)
Comment (jmagman) [reply packages/flutter_tools/test/general.shard/plugins_test.dart:51]: Done.

File: packages/flutter_tools/bin/podhelper.rb (hunk 2 above)
Comment (stuartmorgan) [reply packages/flutter_tools/bin/podhelper.rb:291]: `path_provider` currently doesn't share implementations though, so the macOS implementation package is macOS-only, and it works. 🤔

File: dev/devicelab/bin/tasks/plugin_test_ios.dart [hunk 5]
```diff
@@ -13,6 +13,9 @@ Future<void> main() async {
     // Test that Dart-only plugins are supported.
     PluginTest('ios', <String>['--platforms=ios'], dartOnlyPlugin: true).call,
     PluginTest('macos', <String>['--platforms=macos'], dartOnlyPlugin: true).call,
+    // Test that shared darwin directories are supported.
+    PluginTest('ios', <String>['--platforms=ios,macos'], sharedDarwinSource: true).call,
+    PluginTest('macos', <String>['--platforms=ios,macos'], sharedDarwinSource: true).call,
```
Comment (jmagman) [inline dev/devicelab/bin/tasks/plugin_test_ios.dart:18]: This test is now taking more than 30 minutes.  Will split it up.

Comment (jmagman) [reply dev/devicelab/bin/tasks/plugin_test_ios.dart]: Hm, it's still timing out...

Comment (stuartmorgan) [reply dev/devicelab/bin/tasks/plugin_test_ios.dart]: Looks like the new unit tests I added to the template are working :)

It's probably easiest to ifdef the `result` line that you're already replacing to match the existing implementations than it is to have it rewrite the unit test files as well.

**Comment by jmagman** [comment]
This isn't showing up in the FRoB UI, I don't expect this to cause g3 issues.  Merging.

**Pull request merged**